sympy/matrices/random.py

```diff
@@ -0,0 +1,2343 @@
+    a corresponding $s = \pm \sqrt{1-c^2}$ is choosen randomly.
+
+    Examples
+    ========
+
+    >>> from sympy.core.random import seed
+    >>> from sympy.matrices.random import isometry_normal
+    >>> seed(1)
+
+    >>> isometry_normal(3)
+    Matrix([
+    [ sqrt(2)/2, sqrt(2)/2, 0],
+    [-sqrt(2)/2, sqrt(2)/2, 0],
+    [         0,         0, 1]])
+
+    >>> isometry_normal(3, spec=(0.5, -0.5))
+    Matrix([
+    [               0.5, 0.866025403784439, 0],
+    [-0.866025403784439,               0.5, 0],
+    [                 0,                 0, 1]])
+
+    >>> isometry_normal(3, spec=(-1, 1j, 1))  # no more random
+    Matrix([
+    [-1,     0, 0],
+    [ 0, 1.0*I, 0],
+    [ 0,     0, 1]])
+
+    >>> z = 1+1j
+    >>> isometry_normal(3, spec=(-1, z/abs(z), 1))  # no more random
+    Matrix([
+    [-1,                                       0, 0],
+    [ 0, 0.707106781186547 + 0.707106781186547*I, 0],
+    [ 0,                                       0, 1]])
+
+    Parameters
+    ==========
+    dim : integer
+        dimension n of matrix
+    spec : tuple or list of symbols (optional)
+        set of values of which scalars (diagonal entries) are choosen.
+
+        Each list item has to be
+
+        * either 1 or -1
+        * or any complex unit $z$ (complex number with $|z|=1$)
+        * or a tuple of cosine value $c$ and sine value $s$
+          of a rotation square
+        * or just cosine value $c$ of rotation square (then, a corresponding
+          sin value $s$ will be drawn randomly).
+
+        * If **dim** meets the sum of dimensions of **spec**,
+          spec will be the eigenvalues (diagonal entries) as it is.
+        * If **dim** and the sum of dimensions of **spec** differs,
+          the diagonal entries will be choosen randomly from **spec**.
+        * If not given **spec** defaults to
+          $\{ (\frac{\sqrt(2)}{2}, \frac{\sqrt(2)}{2}), (0, -1) \}$.
+
+    See Also
+    ========
+    rotation
+    diagonal_normal
+    orthogonal
+    unitary
+
+    """
+
+    spec = spec or _rotation_scalars
+
+    # sum of dim of spec
+    _sum = 0
+    for s in spec:
+        _sum += 2 if isinstance(s, (list, tuple)) or not _is_abs_one(s) else 1
+        if dim < _sum:
+            break
+
+    # sample spec
+    if not _sum == dim:
+        spec = [_sample(spec) for _ in range(dim)]
+
+    # set entries of rotation blocks
+    obj = eye(dim)
+    start = 0
+    for s in spec:
+        size, block = 1, s
+        if isinstance(s, (list, tuple)) or not _is_abs_one(s):
+            size, block = 2, rotation(2, (0, 1), s)
+        if dim < start + size:
+            break
+        obj[start, start] = block
+        start += size
+    return obj
+
+
+# === compound matrices, i.e. product of base matrices ===
+
+
+def triangular(dim,
+               rank=None,
+               scalars=_elementary_scalars,
+               units=_elementary_units,
+               length=None):
```

Comment:
I think it would be good to decide that some of these arguments should be keyword-only e.g.:
```python
 def triangular(dim,
￼               rank=None,
                *,
￼               scalars=_elementary_scalars,
￼               units=_elementary_units,
￼               length=None):
```
It's hard to keep track of more than a couple of optional positional arguments.

Reply:
currently the have a common order for _all_ random matrix function: 
```
func(dim, index, rank, spec, scalars, units, length)
```
But which are primarily used depends on the matrix type. But even for `invertible(3, [1,2,3])` a positional argument `scalars` is quite handy.

Reply:
What I mean is that when you see a function call like
```python
f(a,b,c,d,e)
```
It's hard to mentally keep track of what all of the different arguments are supposed to mean (especially if they mean very different things) and what their defaults would be if the arguments are not provided. Calling less common arguments by keyword argument makes it clearer when reading the code:
```python
f(a, b, scalars=c, units=d, length=e)
```
Much of SymPy's codebase dates back to before keyword-only arguments existed:
https://peps.python.org/pep-3102/
Otherwise these would have been used more widely in the codebase.

Reply:
I prefer keyword arguments, too. Turned all non-optional arguments into keyword only arguments.